source/v5/mqtt3_to_mqtt5_adapter.c

```diff
@@ -1651,7 +1681,549 @@ static uint16_t s_aws_mqtt_client_connection_5_publish(
+        };
+        struct aws_mqtt_subscription_set_subscription_record *record =
+            aws_mqtt_subscription_set_subscription_record_new(allocator, &subscription_record_options);
+
+        aws_array_list_push_back(&subscribe_op->subscriptions, &record);
+    }
+
+    /* make temp mqtt5 subscription view array */
+    AWS_VARIABLE_LENGTH_ARRAY(
+        struct aws_mqtt5_subscription_view, mqtt5_subscription_buffer, options->subscription_count);
+    struct aws_mqtt5_subscription_view *subscription_ptr = mqtt5_subscription_buffer;
+    for (size_t i = 0; i < options->subscription_count; ++i) {
+        struct aws_mqtt5_subscription_view *subscription = subscription_ptr + i;
+        AWS_ZERO_STRUCT(*subscription);
+
+        subscription->topic_filter = options->subscriptions[i].topic;
+        subscription->qos = (enum aws_mqtt5_qos)options->subscriptions[i].qos;
+    }
+
+    struct aws_mqtt5_packet_subscribe_view subscribe_view = {
+        .subscriptions = subscription_ptr,
+        .subscription_count = options->subscription_count,
+    };
+
+    struct aws_mqtt5_subscribe_completion_options subscribe_completion_options = {
+        .completion_callback = s_aws_mqtt3_to_mqtt5_adapter_subscribe_completion_fn,
+        .completion_user_data = subscribe_op,
+    };
+
+    subscribe_op->subscribe_op = aws_mqtt5_operation_subscribe_new(
+        allocator, options->adapter->client, &subscribe_view, &subscribe_completion_options);
+
+    if (subscribe_op->subscribe_op == NULL) {
+        goto error;
+    }
+
+    subscribe_op->on_suback = options->on_suback;
+    subscribe_op->on_suback_user_data = options->on_suback_user_data;
+    subscribe_op->on_multi_suback = options->on_multi_suback;
+    subscribe_op->on_multi_suback_user_data = options->on_multi_suback_user_data;
+
+    return subscribe_op;
+
+error:
+
+    aws_mqtt3_to_mqtt5_adapter_operation_release(&subscribe_op->base);
+
+    return NULL;
+}
+
+void s_adapter_subscribe_submission_fn(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+
+    struct aws_mqtt3_to_mqtt5_adapter_operation_subscribe *operation = arg;
+
+    struct aws_mqtt_client_connection_5_impl *adapter = operation->base.adapter;
+
+    size_t subscription_count = aws_array_list_length(&operation->subscriptions);
+    for (size_t i = 0; i < subscription_count; ++i) {
+        struct aws_mqtt_subscription_set_subscription_record *record = NULL;
+        aws_array_list_get_at(&operation->subscriptions, &record, i);
+
+        aws_mqtt_subscription_set_add_subscription(adapter->subscriptions, &record->subscription_view);
+    }
+
+    aws_mqtt5_client_submit_operation_internal(
+        adapter->client, &operation->subscribe_op->base, status != AWS_TASK_STATUS_RUN_READY);
+
+    /* release the transient adapter reference for the move to the event loop */
+    aws_mqtt3_to_mqtt5_adapter_operation_dereference_adapter(&operation->base);
```

Comment:
Trivial: The more I'm seeing it, the stronger I'm feeling about renaming this to `aws_mqtt3_to_mqtt5_adapter_operation_adapter_release` to match what we're using just below with `aws_mqtt3_to_mqtt5_adapter_operation_release`. Haha, actually writing it out though, the repeat use of the word adapter and its extreme similarity to that other release call does also feel weird. I dunno.